ql/src/java/org/apache/hadoop/hive/ql/parse/UpdateDeleteSemanticAnalyzer.java

```diff
@@ -287,6 +299,29 @@ private void reparseAndSuperAnalyze(ASTNode tree, Table mTable, ASTNode tabNameN
     }
   }
 
+  private void checkAndPerformStorageMetadataUpdate(Table table, HiveStorageHandler storageHandler) {
+    if (!deleting() || storageHandler == null) {
+      return;
+    }
+    Map<String, TableScanOperator> topOps = getParseContext().getTopOps();
+    if (!topOps.containsKey(table.getTableName())) {
+      return;
+    }
+    ExprNodeGenericFuncDesc hiveFilter = getParseContext().getTopOps()
+            .get(table.getTableName()).getConf().getFilterExpr();
+    if (hiveFilter == null || !ConvertAstToSearchArg.isCompleteConversion(ctx.getConf(), hiveFilter)) {
```

Comment:
what does it mean isCompleteConversion? no residual?

Reply:
Yes isCompleteConversion means that every leaf node of the hive filter can be converted to a leaf entity in SearchArgument. This ensures that deletes happen the right way and we dont delete anything using partial filters.
hasResiduals also makes sense, let me change it.

Reply:
how about SearchArgument.isPartial()?

Reply:
I think attaching this property to SearchArgument is not a good idea, since it's not a property of the SearchArgument but rather a property of the conversion process itself. Right now I have kept it in ConvertAstToSearchArg itself. If you think this is hard requirement, let me know.

Reply:
The current solution is working. The odd class structure of `ConvertAstToSearchArg` was introduced earlier but I think there is room for improvement since you are touching it anyways:
1. Triggering the parse should be moved from the constructor to a method.

2. Also the class name `ConvertAstToSearchArg` is misleading
* because it is more like a method name.
* we are not converting AST but `ExprNodeDesc` tree

How about `SearchArgumentConverter` ?

3. The result of the conversion should be some result object (like `SearchArgumentConverter.Result`) which wraps both the `SearchArgument` and the boolean property `isPartial`

Reply:
Unfortunately, changing the classname is not compatible since the methods of this class are used by several components. I have created a Result inner class and exposed a new method which provides this Result object which can hold all properties with regards to the conversion process (in this case: partial).